primitives/src/merkle_tree/merkle_tree_impl.rs

```diff
@@ -0,0 +1,413 @@
+                    path.push(branch);
+                    proof.push(MerkleNode::Branch {
+                        value: F::zero(),
+                        children: children
+                            .iter()
+                            .map(|node| {
+                                Box::new(MerkleNode::ForgettenSubtree {
+                                    value: node.value(),
+                                })
+                            })
+                            .collect_vec(),
+                    });
+                    cur = &children[branch];
+                    mpos %= TreeArity::to_u64();
+                    subtree_size /= TreeArity::to_u64();
+                },
+                MerkleNode::EmptySubtree => return LookupResult::EmptyLeaf,
+                MerkleNode::ForgettenSubtree { value: _ } => return LookupResult::NotInMemory,
+            }
+        }
+        LookupResult::Ok((), MerkleProof { pos, path, proof })
+    }
+
+    fn verify(&self, _pos: u64, proof: impl Borrow<Self::Proof>) -> Result<bool, PrimitivesError> {
+        let proof = proof.borrow();
+        let computed_root_value = proof.path.iter().zip(proof.proof.iter()).rev().fold(
+            Ok(F::zero()),
+            |result, (branch, node)| -> Result<F, PrimitivesError> {
+                match result {
+                    Ok(val) => match node {
+                        MerkleNode::Leaf { value: _, children } => Ok(Self::digest_leaf(children)),
+                        MerkleNode::Branch { value: _, children } => {
+                            let mut data = children.iter().map(|node| node.value()).collect_vec();
+                            data[*branch] = val;
+                            Ok(H::digest(&data))
+                        },
+                        _ => Err(PrimitivesError::ParameterError("Invalid proof".to_string())),
+                    },
+                    Err(e) => Err(e),
+                }
+            },
+        )?;
+        Ok(computed_root_value == self.root.value())
+    }
+}
+
+impl<E, H, LeafArity, TreeArity, F> AppendableMerkleTree<F>
+    for MerkleTreeImpl<E, H, LeafArity, TreeArity, F>
+where
+    E: ElementType<F>,
+    H: Hasher<F>,
+    LeafArity: Unsigned,
+    TreeArity: Unsigned,
+    F: Field,
+{
+    fn push(&mut self, elem: &Self::ElementType) -> Result<(), PrimitivesError> {
+        if self.num_leaves == self.capacity {
+            return Err(PrimitivesError::InternalError(
+                "Merkle tree full".to_string(),
+            ));
+        }
+        self.num_leaves += 1;
+        Self::update_internal(
+            &mut self.root,
+            self.height,
+            self.num_leaves - 1,
+            self.capacity,
+            elem,
+        )
+    }
+
+    fn emplace(
+        &mut self,
+        elems: impl Iterator<Item = Self::ElementType>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO(Chengyu): efficient batch insert
```

Comment:
This will be very important for the wallet. We currently make use of [FilledMTBuilder](https://jellyfish.docs.espressosys.com/jf_primitives/merkle_tree/struct.FilledMTBuilder.html) (e.g. [here](https://github.com/EspressoSystems/seahorse/blob/main/src/lw_merkle_tree.rs#L250))

Reply:
Will implement soon.